File: javascript/98. Validate Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

// failed answer
 var min = { val: null }
 var max = { val: null }
 var isValidBST = function(root) {
     let isValid = true;
     let defaultMax = Number.MAX_SAFE_INTEGER
     let defaultMin = Number.MIN_SAFE_INTEGER
     min.val = null
     max.val = null
     function _recursive(root) {
         console.log('-----start------')
         if(root === null) return true;
         console.log('root.val is at', root.val)
         const rightIsValid = _recursive(root.right);
         if(!rightIsValid || (root.right && root.val >= root.right.val) || (min.val && root.val >= min.val)) return false
         const leftIsValid = _recursive(root.left);
         if(!leftIsValid || (root.left && root.val <= root.left.val) || (max.val && root.val <= max.val)) return false
         
         let lastMin = min.val === null ? defaultMax : min.val
         let leftVal = root.left?.val === undefined ? defaultMax : root.left.val
         let rightVal = root.right?.val === undefined ? defaultMax : root.right.val
         min.val = Math.min(lastMin, root.val, leftVal, rightVal)
         let lastMax = max.val === null ? defaultMin : max.val
         leftVal = root.left?.val === undefined ? defaultMin : root.left.val
         rightVal = root.right?.val === undefined ? defaultMin : root.right.val
         max.val = Math.max(lastMax, root.val, leftVal, rightVal)
         console.log('min', min)
         console.log('max', max)
         console.log('leftIsValid', leftIsValid, 'rightIsValid', rightIsValid)
         console.log(`-----end from root.val = ${root.val}------`)
         return leftIsValid && rightIsValid
     }
     
     return _recursive(root)
 };


// correct answer
var isValidBST = function(root) {
  let minSatisfiedValue = Number.MIN_SAFE_INTEGER
  let isValid = true
  function _recursive(root) {
      if(!root) return
      _recursive(root.left)
      if(root.val > minSatisfiedValue) {
          minSatisfiedValue = root.val
      } else {
          isValid = false
      }
      _recursive(root.right)
  }
  
  _recursive(root)
  return isValid;
};

// referenced answer
var isValidBST = function(root, min = Number.MIN_SAFE_INTEGER, max = Number.MAX_SAFE_INTEGER) {
  if(!root) return true
  if(root.val < min || root.val > max) return false

  return isValidBST(root.left, min, root.val) && isValidBST(root.right, root.val, max)
}

File: golang/trees/98. Validate Binary Search Tree.go
package trees

import (
	"math"
)

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// failed
//func isValidBST(root *TreeNode) bool {
//	// check each node if it's valid BST
//	// check left subtree
//	// check right subtree
//	var dfs func(root *TreeNode, a int, b int) *info
//	dfs = func(root *TreeNode, a int, b int) *info {
//		if root == nil {
//			return &info{a, b, true}
//		}
//
//		lInfo := dfs(root.Left, a, b)
//		rInfo := dfs(root.Right, a, b)
//
//		if root.Val < lInfo.Max || root.Val > rInfo.Min {
//			return &info{min(lInfo.Min, rInfo.Min, root.Val), max(lInfo.Max, rInfo.Max, b), false}
//		}
//
//		return &info{min(root.Val, lInfo.Min, rInfo.Min), max(root.Val, lInfo.Max, rInfo.Max), lInfo.Valid && rInfo.Valid}
//	}
//
//	ans := dfs(root, math.MaxInt16, math.MinInt16)
//	return ans.Valid
//}

func isValidBST(root *TreeNode) bool {
	if root == nil {
		return true
	}

	if !isValid(root.Left, root.Val, func(val, limit int) bool { return val < limit }) ||
		!isValid(root.Right, root.Val, func(val, limit int) bool { return val > limit }) {
		return false
	}

	return isValidBST(root.Left) && isValidBST(root.Right)
}

func isValid(root *TreeNode, limit int, check func(a int, b int) bool) bool {
	if root == nil {
		return true
	}
	if !check(root.Val, limit) {
		return false
	}
	return isValid(root.Left, limit, check) && isValid(root.Right, limit, check)
}

func isValidBST2(root *TreeNode) bool {
	return isValid2(root, math.MinInt64, math.MaxInt64)
}

func isValid2(root *TreeNode, lower, upper int) bool {
	if root == nil {
		return true
	}

	if root.Val < lower || root.Val > upper {
		return false
	}

	return isValid2(root.Left, lower, root.Val) && isValid2(root.Right, root.Val, upper)
}

func isValidBST3(root *TreeNode) bool {
	return valid(root, math.MinInt64, math.MaxInt64)
}

func valid(root *TreeNode, minVal, maxVal int) bool {
	if root == nil {
		return true
	}

	if root.Val < minVal || root.Val > maxVal {
		return false
	}

	return valid(root.Left, minVal, root.Val) && valid(root.Right, root.Val, maxVal)
}
